iphone/cli/lib/process-js-task.js

```diff
@@ -0,0 +1,374 @@
+	/**
+	 * Constructs a new processing task.
+	 *
+	 * @param {Object} taskInfo Configuration object for this task
+	 */
+	constructor(taskInfo) {
+		taskInfo.name = taskInfo.name || 'process-js';
+		super(taskInfo);
+
+		this.jsFiles = taskInfo.jsFiles;
+		this.jsBootstrapFiles = taskInfo.jsBootstrapFiles;
+		this.builder = taskInfo.builder;
+		this.sdkCOmmonFolder = taskInfo.sdkCOmmonFolder;
+
+		const minify = this.builder.minifyJS;
+		const transpile = this.builder.transpile;
+		this.defaultAnalyzeOptions = {
+			minify,
+			transpile,
+			sourceMap: this.builder.sourceMaps || this.builder.deployType === 'development',
+			resourcesDir: this.builder.xcodeAppDir,
+			logger: this.logger,
+			targets: {
+				ios: this.builder.minSupportedIosSdk
+			}
+		};
+
+		this.dataFilePath = path.join(this.incrementalDirectory, 'data.json');
+		this.resetTaskData();
+	}
+
+	/**
+	 * Does a full task run, processing every input file.
+	 *
+	 * @return {Promise}
+	 */
+	doFullTaskRun() {
+		this.resetTaskData();
+		return Promise.all(Array.from(this.inputFiles).map(filePath => this.processJsFile(filePath)));
+	}
+
+	/**
+	 * Does an incremental task run, processing only changed files.
+	 *
+	 * For backwards compatibiliy this will currently also call processJsFile for unchanged files.
+	 *
+	 * @param {Map} changedFiles Map of file paths and their current file state (created, changed, deleted)
+	 * @return {Promise}
+	 */
+	doIncrementalTaskRun(changedFiles) {
+		const loaded = this.loadTaskData();
+		const fullBuild = !loaded || this.requiresFullBuild();
+		if (fullBuild) {
+			return this.doFullTaskRun();
+		}
+
+		this.jsBootstrapFiles.splice(0, 0, ...this.data.jsBootstrapFiles);
+
+		const deletedFiles = this.filterFilesByStatus(changedFiles, 'deleted');
+		const deletedPromise = Promise.all(deletedFiles.map(filePath => this.handleDeletedFile(filePath)));
+
+		const updatedFiles = this.filterFilesByStatus(changedFiles, [ 'created', 'changed' ]);
+		const updatedPromise = Promise.all(updatedFiles.map(filePath => this.processJsFile(filePath)));
+
+		// @fixme: can be removed in 9.0 to even further decrease build times on incremental builds
+		const unchangedFiles = Array.from(this.inputFiles).filter(filePath => !changedFiles.has(filePath));
+		const unchangedPromise = Promise.all(unchangedFiles.map(filePath => this.processJsFile(filePath)));
+
+		return Promise.all([ deletedPromise, updatedPromise, unchangedPromise ]);
+	}
+
+	/**
+	 * Loads last run's task data if this task will be skipped entirely.
+	 *
+	 * Note that this function will be called when no input files changed. Since
+	 * we also depend on some config values from the builder this is used to
+	 * fallback to a full task run if required.
+	 *
+	 * @return {Promise}
+	 */
+	loadResultAndSkip() {
+		const loaded = this.loadTaskData();
+		const fullBuild = !loaded || this.requiresFullBuild();
+		if (fullBuild) {
+			return this.doFullTaskRun();
+		}
+
+		return Promise.resolve();
```

Comment:
This seems self-defeating. Why not make the method async so it just does the right thing? Seems to me like in the non full build scenario it will run the first two lines sync and then return an empty Promise.

```suggestion
	async loadResultAndSkip() {
		const loaded = await this.loadTaskData();
		const fullBuild = !loaded || this.requiresFullBuild();
		if (fullBuild) {
			return this.doFullTaskRun();
		}
	}
```